### PR TITLE
NN Archive - Parsers Support

### PR DESCRIPTION
**Pull request by kkeroo**
This PR adds support for parsers for all heads. Now, each head has `parser_name` and `metadata` that stores key-value pairs for required parameters. Parameters from the old head are moved to `metadata` object (e.g. `n_keypoints`, `classes`,...).

Each head in NN archive has its own `Metadata` model. Starting with the basic head having `HeadMetadata`, then `HeadObjectDetection` having `HeadObjectDetectionMetadata`. All `Metadata` models also allow you to have custom fields, but they still validate the required ones. The only problem here is that code editor still underlines the custom parameter as unsupported, you can avoid this by adding `#type: ignore`. Does anyone know a better way?
```
metadata = HeadMetadata(
    classes=['person', 'car'],
    n_classes=2,
    custom_parameter='custom', # type: ignore
)
```

YOLO heads still check the right combination of parameters and allow multiple tasks.

Additionally, one test is added for testing the custom parameter setting in `Metadata` object.

One example of `config.json` with custom parameter:
```
{
    "config_version": "1.0",
    "model": {
        "metadata": {
            "name": "XFeat",
            "path": "matching.onnx",
            "precision": "float32"
        },
        "inputs": [
            {
                "name": "images",
                "dtype": "float32",
                "input_type": "image",
                "shape": [
                    1,
                    3,
                    352,
                    640
                ],
                "layout": "NCHW",
                "preprocessing": {
                    "mean": [
                        0.0,
                        0.0,
                        0.0
                    ],
                    "scale": [
                        255.0,
                        255.0,
                        255.0
                    ],
                    "reverse_channels": false,
                    "interleaved_to_planar": false
                }
            }
        ],
        "outputs": [
            {
                "name": "output",
                "dtype": "float32"
            }
        ],
        "heads": [
            {
                "family": "Segmentation",
                "parser_name": "PP-lites",
                "metadata": {
                    "classes": [
                        "person",
                        "car"
                    ],
                    "n_classes": 2,
                    "is_softmax": false,
                    "custom_field": "custom_field"
                },
                "outputs": {
                    "predictions": "output"
                }
            }
        ]
    }
}
```

### CODE REVIEW

**Comment by kkeroo** [comment]
Dataset test is failing, this has nothing to do with the changes.

File: luxonis_ml/nn_archive/config_building_blocks/base_models/head.py [hunk 1]
```diff
@@ -144,6 +163,15 @@ def validate_label_type(
             raise ValueError("Invalid family")
         return value
 
+    @field_validator("metadata")
+    def validate_metadata(cls, value):
+        if not isinstance(value, HeadSegmentationMetadata):
```
Comment (kozlov721) [inline luxonis_ml/nn_archive/config_building_blocks/base_models/head.py:168]: Can this ever happen with the `pydantic` validation already in place?

File: luxonis_ml/nn_archive/config_building_blocks/base_models/head.py [hunk 2]
```diff
@@ -20,51 +27,42 @@ class Head(BaseModel, ABC):
...
-    classes: List[str] = Field(
-        description="Names of object classes recognized by the model."
-    )
-    n_classes: int = Field(
-        description="Number of object classes recognized by the model."
-    )
+
+    @field_validator("metadata")
+    def validate_metadata(cls, value):
+        if not isinstance(value, HeadMetadata):
```
Comment (kozlov721) [inline luxonis_ml/nn_archive/config_building_blocks/base_models/head.py:43]: Can this ever happen with the `pydantic` validation already in place?

File: luxonis_ml/nn_archive/config_building_blocks/base_models/head.py [hunk 3]
```diff
@@ -93,6 +91,15 @@ def validate_label_type(
             raise ValueError("Invalid family")
         return value
 
+    @field_validator("metadata")
+    def validate_metadata(cls, value):
+        if not isinstance(value, HeadClassificationMetadata):
```
Comment (kozlov721) [inline luxonis_ml/nn_archive/config_building_blocks/base_models/head.py:96]: Same as above

File: luxonis_ml/nn_archive/config_building_blocks/base_models/head.py [hunk 4]
```diff
@@ -117,6 +127,15 @@ def validate_label_type(
             raise ValueError("Invalid family")
         return value
 
+    @field_validator("metadata")
+    def validate_metadata(cls, value):
+        if not isinstance(value, HeadObjectDetectionMetadata):
```
Comment (kozlov721) [inline luxonis_ml/nn_archive/config_building_blocks/base_models/head.py:132]: Same as above

File: luxonis_ml/nn_archive/config_building_blocks/base_models/head.py [hunk 5]
```diff
@@ -195,12 +199,23 @@ def validate_label_type(
             raise ValueError("Invalid family")
         return value
 
+    @field_validator("metadata")
+    def validate_metadata(cls, value):
+        if not isinstance(value, HeadYOLOMetadata):
```
Comment (kozlov721) [inline luxonis_ml/nn_archive/config_building_blocks/base_models/head.py:204]: Same as above

File: luxonis_ml/nn_archive/config_building_blocks/base_models/head.py (hunk 1 above)
Comment (kkeroo) [reply luxonis_ml/nn_archive/config_building_blocks/base_models/head.py]: Thanks, removed in [1031e73](https://github.com/luxonis/luxonis-ml/pull/127/commits/1031e733652d54e98130f6baf7400c621414b5eb).

File: luxonis_ml/nn_archive/config_building_blocks/base_models/head.py (hunk 2 above)
Comment (kkeroo) [reply luxonis_ml/nn_archive/config_building_blocks/base_models/head.py]: Thanks, removed in [1031e73](https://github.com/luxonis/luxonis-ml/pull/127/commits/1031e733652d54e98130f6baf7400c621414b5eb).

File: luxonis_ml/nn_archive/config_building_blocks/base_models/head.py (hunk 3 above)
Comment (kkeroo) [reply luxonis_ml/nn_archive/config_building_blocks/base_models/head.py]: Thanks, removed in [1031e73](https://github.com/luxonis/luxonis-ml/pull/127/commits/1031e733652d54e98130f6baf7400c621414b5eb).

File: luxonis_ml/nn_archive/config_building_blocks/base_models/head.py (hunk 4 above)
Comment (kkeroo) [reply luxonis_ml/nn_archive/config_building_blocks/base_models/head.py]: Thanks, removed in [1031e73](https://github.com/luxonis/luxonis-ml/pull/127/commits/1031e733652d54e98130f6baf7400c621414b5eb).

File: luxonis_ml/nn_archive/config_building_blocks/base_models/head.py (hunk 5 above)
Comment (kkeroo) [reply luxonis_ml/nn_archive/config_building_blocks/base_models/head.py]: Thanks, removed in [1031e73](https://github.com/luxonis/luxonis-ml/pull/127/commits/1031e733652d54e98130f6baf7400c621414b5eb).

**Review by kozlov721 — APPROVED** [review]
LGTM

**Comment by kkeroo** [comment]
### Update

We removed `family` from all heads and replaced it with `parser`. Also, `postprocessor_path` field is in every head and removed from Yolo's head metadata.

Every head now looks like:
```
Head:
  parser
  postprocessor_path
  metadata
  outputs
```

**Review by tersekmatija — APPROVED** [review]
Looks good, let's make the following optional on the Inputs so we can deprecate it inside DAI and make more explicit later:
```
                    "reverse_channels": true,
                    "interleaved_to_planar": false
```

**Comment by kkeroo** [comment]
### Update

1. `postprocesor_path` back in `Metadata`
2. Optional `reverse_input_channels` and `interleaved_to_planar`
3. `outputs` on the head level only lists of strings
4. `outputs` in `Metadata` only for `YOLO` and `SSD`
5. validation moved to `Metadata`
6. only a single `Head`

**Review by tersekmatija — APPROVED** [review]

**Pull request merged**